sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RecoverableReactorConnection.java

```diff
@@ -0,0 +1,249 @@
+     * @param fullyQualifiedNamespace The connection FQDN of the remote broker/resource.
+     * @param entityPath The relative path to the entity under the FQDN to which the connection established to.
+     * @param retryPolicy the retry configuration to use to obtain a new active connection.
+     * @param loggingContext the logger context.
+     */
+    public RecoverableReactorConnection(Supplier<C> connectionSupplier,
+                                        String fullyQualifiedNamespace,
+                                        String entityPath,
+                                        AmqpRetryPolicy retryPolicy,
+                                        Map<String, Object> loggingContext) {
+        this.fullyQualifiedNamespace = Objects.requireNonNull(fullyQualifiedNamespace, "'fullyQualifiedNamespace' cannot be null.");
+        // Note: FQDN, (to an extent) entity-path are generic enough, but if we find more connection description
+        // parameters that are non-generic, i.e., specific to individual messaging services, then consider creating
+        // dedicated POJO types to pass around connection description parameters in corresponding libraries rather
+        // than polluting shared 'RecoverableReactorConnection' type.
+        this.entityPath = entityPath;
+        Objects.requireNonNull(retryPolicy, "'retryPolicy' cannot be null.");
+        this.retryOptions = retryPolicy.getRetryOptions();
+        this.errorContext = new AmqpErrorContext(fullyQualifiedNamespace);
+        this.logger = new ClientLogger(getClass(), Objects.requireNonNull(loggingContext, "'loggingContext' cannot be null."));
+        Objects.requireNonNull(connectionSupplier, "'connectionSupplier' cannot be null.");
+        final Mono<C> newConnection = Mono.fromSupplier(() -> {
+            if (terminated) {
+                logger.info("Connection recovery support is terminated, dropping the request for new connection.");
+                throw TERMINATED_ERROR;
+            } else {
+                return connectionSupplier.get();
+            }
+        });
+
+        this.createOrGetCachedConnection = newConnection
+            .flatMap(c -> {
+                logger.atInfo()
+                    .addKeyValue(CONNECTION_ID_KEY, c.getId())
+                    .log("Waiting to connect and become active.");
+
+                return c.connectAndAwaitToActive()
+                    .doOnCancel(() -> {
+                        if (!c.isDisposed()) {
+                            c.closeAsync(closeSignal("The connection request was canceled while waiting to active."))
+                                .subscribe();
+                        }
+                    });
+            })
+            .retryWhen(retryWhenSpec(retryPolicy))
+            .<C>handle((c, sink) -> {
+                @SuppressWarnings("unchecked")
+                final C connection = (C) c;
+                currentConnection = connection;
```

Comment:
There is no read-write lock on `currentConnection`, is it possible that multiple threads write to this variable? (I noticed the flatMap operator up there) because we also read from it in dispose()

Reply:
Two blocks of code that can execute concurrently is -

<table>
<tr>
<td> Block1 </td> <td> Block2 </td>
</tr>
<tr>
<td>

```java
this.currentConnection = connection;
 if (this.terminated) {
    connection.closeAsync(...);
 }
```

</td>
<td>

```java
this.terminated = true;
this.currentConnection.closeAsync(...);
```

</td>
</tr>
</table>

Both 'currentConnection' and 'terminated' are volatile; hence, 

1. Write to 'currentConnection' in block1 is immediately visible to  its read in block2.
2. Write to 'terminated' in block2 is immediately visible to its read in block1.

If we look at the read-write order of these volatile variables combined with JVM guarantee that there won't be any reordering, my understanding is it is impossible to leak a connection. But please have a look and see if I overlooked it.

Reply:
Actually, let me get back on this after doing digging the spec

Reply:
Digging the spec, I don't see an explicit statement asserting that JVM won't reorder the above volatile r-w. Given these two blocks are almost never contended and are not hot-path, I would go with _synchronized_.

Reply:
Updated to use _synchronized_